pkg/controller/keyvault/spc/gateway.go

```diff
@@ -0,0 +1,204 @@
+)
+
+var gatewaySecretProviderControllerName = controllername.New("gateway", "keyvault", "secret", "provider")
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config, serviceAccountIndexName string) error {
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+	if conf.DisableKeyvault {
+		return nil
+	}
+
+	spcReconciler := &secretProviderClassReconciler[*gatewayv1.Gateway]{
+		name: gatewaySecretProviderControllerName,
+		toSpcOpts: func(ctx context.Context, cl client.Client, gw *gatewayv1.Gateway) iter.Seq2[spcOpts, error] {
+			return gatewayToSpcOpts(ctx, cl, conf, gw)
+		},
+
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	}
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}).
+			Owns(&secv1.SecretProviderClass{}).
+			Watches(&corev1.ServiceAccount{}, handler.EnqueueRequestsFromMapFunc(util.GenerateGatewayGetter(manager, serviceAccountIndexName))),
+		manager.GetLogger(),
+	).Complete(spcReconciler)
+}
+
+func gatewayToSpcOpts(ctx context.Context, cl client.Client, conf *config.Config, gw *gatewayv1.Gateway) iter.Seq2[spcOpts, error] {
+	return func(yield func(spcOpts, error) bool) {
+		if conf == nil {
+			yield(spcOpts{}, errors.New("config is nil"))
+			return
+		}
+
+		if gw == nil {
+			yield(spcOpts{}, errors.New("gateway is nil"))
+			return
+		}
+
+		if !IsManagedGateway(gw) {
+			// todo: test this and make sure it returns no values instead of hangs
+			return
+		}
+
+		for index, listener := range gw.Spec.Listeners {
+			name := GetGatewayListenerSpcName(gw.Name, string(listener.Name))
+			opts := spcOpts{
+				action:     actionReconcile,
+				name:       name,
+				namespace:  gw.Namespace,
+				tenantId:   conf.TenantID,
+				secretName: name,
+				cloud:      conf.Cloud,
+			}
+
+			if !ListenerIsKvEnabled(listener) {
+				opts.action = actionCleanup
+				if !yield(opts, nil) {
+					return
+				}
+				continue
+			}
+
+			clientId, err := clientIdFromListener(ctx, cl, gw.Namespace, listener)
+			if err != nil {
+				if !yield(opts, err) {
+					return
+				}
+				continue
+			}
+			opts.clientId = clientId
+
+			uri := string(listener.TLS.Options[certUriTLSOption])
+			certRef, err := parseKeyVaultCertURI(uri)
+			if err != nil {
+				if !yield(opts, fmt.Errorf("parsing KeyVault cert URI %s: %w", uri, err)) {
+					return
+				}
+				continue
+			}
+
+			opts.vaultName = certRef.vaultName
+			opts.certName = certRef.certName
+			opts.objectVersion = certRef.objectVersion
+			opts.modifyOwner = func(obj client.Object) error {
+				gwObj, ok := obj.(*gatewayv1.Gateway)
+				if !ok {
+					return fmt.Errorf("object is not a Gateway: %T", obj)
+				}
+
+				gwObj.Spec.Listeners[index].TLS.CertificateRefs = []gatewayv1.SecretObjectReference{
+					{
+						Namespace: util.ToPtr(gatewayv1.Namespace(opts.namespace)),
```

Comment:
why not use gw namespace?

Reply:
opts.namespace is guaranteed to be where the spc is

discussed on teams